### PR TITLE
Improve ZeRO Partitioning Algorithm

### PR DESCRIPTION
**Pull request by awgu**
Task: https://www.internalfb.com/tasks/?t=90847711

**Overview:**
Suppose we have `n` items with positive integer sizes and `k` buckets. We want to assign items to buckets with the goal of uniformity. The precise criteria for uniformity can vary: e.g. minimize the maximum size, maximize the minimum size, etc. This is known as [multiway number partitioning](https://en.wikipedia.org/wiki/Multiway_number_partitioning). ZeRO's partitioning task reduces to solving this problem. In particular, this is the subproblem to be solved for each `param_group` in `self.param_groups`, where the parameters are the items and the ranks give the buckets.

The existing implementation uses the linear-time [greedy number partitioning algorithm](https://en.wikipedia.org/wiki/Greedy_number_partitioning#Linear-time_algorithm), which assigns the next tensor-parameter to the process with the smallest total parameter size so far. In this task, I explore the [extension](https://en.wikipedia.org/wiki/Greedy_number_partitioning#Improved_algorithm) where each parameter group is sorted by decreasing size before applying the greedy algorithm, requiring linearithmic time (as dominated by the sort). 

**Experiments**
The mean number of parameters represents a perfectly uniform allocation and hence the ideal allocation (which may be even better than the optimal partition). In the following tables, I present the maximum number of parameters for any one process and the difference from the mean in parentheses for ResNet-50, ResNet-152, and BERT (the bare BERT model). The best-performing partitioning strategy for each model is bolded.

Two processes:
| Model | Max Num Params - Greedy (Diff) | Max Num Params - Greedy-Sorted (Diff) | Mean Num Params |
| --- | --- | --- | --- |
| ResNet-50 | 13,249,600 (471,084) | **12,794,816 (16,300)** | 12,778,516 |
| ResNet-152 | 30,567,488 (471,084) | **30,111,424 (15,020)** | 30,096,404 |
| BERT | **54,749,184 (8,064)** | 55,327,488 (586,368) | 54,741,120 |

Four processes:
| Model | Max Num Params - Greedy (Diff) | Max Num Params - Greedy-Sorted (Diff) | Mean Num Params |
| --- | --- | --- | --- |
| ResNet-50 | 7,524,864 (1,135,606) |  **6,436,864 (47,606)** | 6,389,258 |
| ResNet-152 | 16,232,192 (1,183,990) | **15,090,152 (41,950)** | 15,048,202 |
| BERT | **28,151,040 (780,480)** | 28,352,256 (981,696)  | 27,370,560 |

---

I also investigated the latency of `optimizer.step()` for the different partitioning algorithms. I measured the latency for 30 iterations and took the mean latency per process (excluding the first iteration due to cache coldness). In the following tables, I present the maximum of those mean latencies over all processes and the standard deviation of the latencies contributing to that maximum. Again, the best-performing partitioning strategy for each model is bolded. All entries are presented in seconds and used `gloo` backend.

Two processes:
| Model | Max `optimizer.step()` Time - Greedy (Std.) | Max `optimizer.step()` Time - Greedy-Sorted (Std.) |
| --- | --- | --- |
| ResNet-50 | **0.060 (0.002)** | 0.061 (0.002) |
| ResNet-152 | 0.166 (0.003) | **0.160 (0.004)** |
| BERT | 0.220 (0.009) | **0.199 (0.006)** |

Four processes:
| Model | Max `optimizer.step()` Time - Greedy | Max `optimizer.step()` Time - Greedy-Sorted |
| --- | --- | --- |
| ResNet-50 | 0.094 (0.004) | **0.093 (0.004)** |
| ResNet-152 | **0.228 (0.011)** | 0.231 (0.009) |
| BERT | **0.328 (0.015)** | 0.329 (0.021) |

Based on the standard deviations, the differences in the latency measurements across the different algorithms appear to be within the uncertainty in the measurement itself. Hence, it is difficult to argue that one algorithm is clearly the fastest.

---

`zero.py` is my experiment script, and I use the AI AWS cluster. The run command looks like: 
```
srun -p $DEV_QUEUE --cpus-per-task=16 -t 5:00:00 --gpus-per-node=4 python zero.py -b nccl greedy 2 4
```
This runs the experiment script on an instance with 4 GPUs using `nccl` backend, outputting to a directory named `greedy/`, and using world sizes of 2 and 4. An analogous command can be used after modifying `partition_parameters()`, e.g. replacing `greedy` with `greedy_sorted` as the output directory name. Then, to run the analysis script:
```
python analyze.py greedy greedy_sorted
```

**Notes:**
There exists an optimal solution to this partitioning problem. An algorithm that finds such a solution is the [complete greedy algorithm (CGA)](https://en.wikipedia.org/wiki/Greedy_number_partitioning#An_exact_algorithm), which reduces to the brute-force combinatorial search in the worst case. There exist heuristics to improve the `k = 2` case (i.e. when there are two processes); however, given that `n` in typical use cases is very large, any algorithm that is quadratic or slower is unrealistic. Other exact algorithms are similarly exponential in the worst case, rendering them intractable.

Given this, I do not currently see a need for future proofing the partitioning algorithm against the introduction of algorithms beyond the naive greedy and the sorted greedy algorithms. The flexibility I can imagine though is possibly specifying between the two (perhaps, leaving one as the default). Another idea I had is that because this partitioning only happens upon initialization, we could make multiple passes over the parameters, partitioning them in several ways and taking the best partition. 
**Note:** As of right now (end of June 4), the code change simply adapts the sorted greedy algorithm with no attempts at modularization. I can change the code accordingly after we decide what level of flexibility should be offered.

---

In the current ZeRO implementation, the core `partition_parameters()` computation happens twice upon initialization (i.e. call to `__init__()`): first from a call to `_param_to_rank()` (i.e. an access to `_param_to_rank`) and then from a call to `_update_trainable()`. `_update_trainable()` sees that no optimizer has been constructed yet, so it clears the cache, eliminating the first `partition_parameters()` computation and performing a redundant re-computation. 

Here is a typical trace:
- [The ZeRO optimizer object is initialized, calling `__init__()`.](https://github.com/pytorch/pytorch/blob/d125694d0bc4e02de9a54ce485b31ca333559203/torch/distributed/optim/zero_redundancy_optimizer.py#L142)
- [In `__init__()`, `self._device` is set, so it accesses `self._per_device_params`.](https://github.com/pytorch/pytorch/blob/d125694d0bc4e02de9a54ce485b31ca333559203/torch/distributed/optim/zero_redundancy_optimizer.py#L182)
- [`self._per_device_params` is not cached, so it accesses `self._param_to_rank`.](https://github.com/pytorch/pytorch/blob/d125694d0bc4e02de9a54ce485b31ca333559203/torch/distributed/optim/zero_redundancy_optimizer.py#L340)
- [`self._param_to_rank` is not cached, so it calls `partition_parameters()`.](https://github.com/pytorch/pytorch/blob/d125694d0bc4e02de9a54ce485b31ca333559203/torch/distributed/optim/zero_redundancy_optimizer.py#L353) (first call to `partition_parameters()`)
- [`__init__()` later calls `_update_trainable()`.](https://github.com/pytorch/pytorch/blob/d125694d0bc4e02de9a54ce485b31ca333559203/torch/distributed/optim/zero_redundancy_optimizer.py#L185)
- [In `_update_trainable()`, `self` does not have `attr` `"optim"`, so it clears the cached objects (notably, `self._partition_parameters_cache`).](https://github.com/pytorch/pytorch/blob/d125694d0bc4e02de9a54ce485b31ca333559203/torch/distributed/optim/zero_redundancy_optimizer.py#L591)
- [`_update_trainable()` calls `self.partition_parameters()`.](https://github.com/pytorch/pytorch/blob/d125694d0bc4e02de9a54ce485b31ca333559203/torch/distributed/optim/zero_redundancy_optimizer.py#L593) (second call to `partition_parameters()`)

### CODE REVIEW

**Review by mrshenli — APPROVED** [review]
Thanks a lot for conducting comprehensive experiments. Given that this new algorithm considerably improves the result for ResNet and only slightly degrades for BERT, I'd vote for landing this PR.

**Comment by mrshenli** [comment]
> In the current ZeRO implementation, the core partition_parameters() computation happens twice upon initialization (i.e. call to __init__()): 

@blefaudeux is this intentional? Is not, we can work on a followup PR to consolidate the two into one.

**Comment by blefaudeux** [comment]
> > In the current ZeRO implementation, the core partition_parameters() computation happens twice upon initialization (i.e. call to **init**()):
> 
> @blefaudeux is this intentional? Is not, we can work on a followup PR to consolidate the two into one.

no it's not intentional, mistake indeed, good catch !

**Comment by blefaudeux** [comment]
Thanks for the very comprehensive write up @andwgu ! For completeness, I'll add a consideration to the partitioning: something else which could be done, and which I think could be meaningful speed-wise, is to make the partitioning order-aware, and DDP non-blocking on the reduce part in that case (each rank could just check that its own parameters are reduced before stepping, the lock being moved to after the step/broadcast). It's a much, much bigger change but one which allow for some overlap in between the backward and the step. Just food for thoughts, I think that we may have discussed that somewhere else already

**Comment by mrshenli** [comment]
Hey @andwgu, shall we move the benchmark code into a separate PR? If you need to have that PR sitting on top of this PR, you can use the ghstack tool: https://github.com/ezyang/ghstack

Thanks!

**Comment by pritamdamania87** [comment]
> Thanks a lot for conducting comprehensive experiments. Given that this new algorithm considerably improves the result for ResNet and only slightly degrades for BERT, I'd vote for landing this PR.

What is the high level metric we're trying to optimize for here? For example the optimizer.step() latency doesn't change much based on these changes. If we're looking at better load balancing of parameters for GPU memory utilization, we should probably track that metric and see if there is substantial improvement in balanced GPU memory. Also, I am concerned about the results for BERT since Transformer models are actually the number one use case where techniques like ZeRO optimizer would be used.

**Comment by pritamdamania87** [comment]
> however, given that n in typical use cases is very large, any algorithm that is quadratic or slower is unrealistic.

How much is n for the models we compared here? My understanding is that `n` here is not actually the total number of parameters in the model but `len(model.parameters())` which is probably much lower? I'm wondering if this algorithm is feasible: https://en.wikipedia.org/wiki/Multiway_number_partitioning#Reduction_to_bin_packing?

**Comment by andwgu** [comment]
> How much is n for the models we compared here? My understanding is that `n` here is not actually the total number of parameters in the model but `len(model.parameters())` which is probably much lower? I'm wondering if this algorithm is feasible: https://en.wikipedia.org/wiki/Multiway_number_partitioning#Reduction_to_bin_packing?

You are right, `n` is not the total number of parameters in the model but rather `len(model.parameters())`. My fault for any confusion. `len(model.parameters())` (i.e. `len(param_group["params"]` assuming one parameter group) for each of the three models are as follows:
- ResNet-50: 161
- ResNet-152: 467
- BERT: 199

Given the magnitude of these numbers, my intuition is that we can relax the time complexity requirement, but the constant factors will matter. We should also be clear on when this partitioning takes place: Right now, this is done once in the constructor and also any time that the parameter trainability changes (e.g. if some parameters are frozen or unfrozen). If we do not anticipate the parameter trainability to change often, then a more computationally intensive partition algorithm can be considered. Alternatively, we can use a heavier-weight algorithm at construction time and use faster algorithms thereafter, or we can allow the user to specify the algorithm(s) from a fixed list of choices (I say _fixed_ since I think that user-defined algorithms may be challenging to integrate).

I agree that we should establish what metric we are trying to optimize. Initially, I assumed the goodness of the partition would directly translate to an actual performance-impacting metric, but now I see that this may not be the case. As you mentioned, the optimizer `step()` latency is roughly the same. I am not experienced in the collective communications/networks, but perhaps, the usage of `broadcast()` in the `step()` implementation poses a limit? Each rank broadcasts its shard of the parameters asynchronously, but the degree to which a more-uniform partition helps depends on how much the broadcasts are overlapped.

Either way, I took a look at the linked algorithm. Denote multiway number partitioning (our problem) by MNP and bin packing by BP. To provide a brief summary: In both MNP and BP, we have `n` positive integers `S = {s_1, ..., s_n}`. In MNP, we want to assign them to `k` bins so that the largest bin is minimized. In BP, we have a capacity `C` and want to assign them into the minimum number of bins so that no bin exceeds `C`. The algorithm (solving MNP using a BP oracle) starts with a cleverly-chosen lower and upper bound on `C` and binary searches for the minimum `C` such that the `n` integers can be packed into at most `k` bins.

My concern is that finding the lower and upper bounds each take at least one pass over the parameters (or more precisely, at least `n + n log n` work for even the simplest algorithms). Then, if the difference in the bounds is `m`, there are another `log m` calls to the bin packing oracle, which itself requires at least one pass over the parameters. This results in several more passes over the parameters compared to the simple greedy algorithm or the sorted-greedy algorithm. Again, it is difficult to evaluate this tradeoff without first identifying the metric to optimize. 

If we decide that GPU memory utilization is the metric to track, let me know, and I can rework the experiment and update the results.

**Comment by andwgu** [comment]
Here, I present the results of the [new partitioning algorithm](https://en.wikipedia.org/wiki/Multiway_number_partitioning#Reduction_to_bin_packing) suggested by @pritamdamania.

**TL;DR: I incorrectly implemented the sorted-greedy algorithm previously, and the new algorithm performs no better than the sorted-greedy algorithm.**

**Algorithm:**
ZeRO's partitioning is an instance of the multiway number partitioning (MNP) problem: Suppose we have `n` positive integers `s_1, ..., s_n` and `k` bins. We want to pack elements into bins so that the largest bin is minimized. A dual to MNP is the bin packing (BP) problem: Given a max capacity `C` per bin, we want to pack the `n` elements into as few bins as possible. The new algorithm (sometimes known as multifit) leverages the duality by binary searching for the minimum capacity `C` such that the `n` elements can be packed into `k` or fewer bins. 

In my implementation, I lower bound the search space for `C` by `(sum_{i=1}^{n} s_i) / k` and upper bound it by `min{max bin size from greedy, max bin size from sorted-greedy}`. This upper bound implies that the new algorithm cannot do worse than the greedy and sorted-greedy algorithms. However, we shall see below that on the three considered models, the algorithm **only matches** the sorted-greedy algorithm's performance.

**Disclaimer:**
In the process of implementing this new algorithm, I caught a crucial mistake in my implementation in the sorted-greedy algorithm. I was sorting by the first dimension of each tensor's size instead of the number of elements in the tensor:
```
# Old incorrect version:
params_sorted = sorted(param_group["params"], key=lambda t: t.size()[0], reverse=True)
# Corrected version:
params_sorted = sorted(param_group["params"], key=lambda t: t.numel(), reverse=True)
```

**Updated Results:**
Two processes:
| Model | Max Num Params - Greedy (Diff) | Max Num Params - Sorted-Greedy (Diff) | Max Num Params - Multifit (Diff) | Mean Num Params |
| --- | --- | --- | --- | --- |
| ResNet-50 | 13,249,600 (471,084) | **12,778,536 (20)** | **12,778,536 (20)** | 12,778,516 |
| ResNet-152 | 30,567,488 (471,084) | **30,096,424 (20)** | **30,096,424 (20)** | 30,096,404 |
| BERT | 54,749,184 (8,064) | **54,741,504 (384)** | **54,741,504 (384)** | 54,741,120 |

Four processes:
| Model | Max Num Params - Greedy (Diff) | Max Num Params - Sorted-Greedy (Diff) | Max Num Params - Multifit (Diff) | Mean Num Params |
| --- | --- | --- | --- | --- |
| ResNet-50 | 7,524,864 (1,135,606) |  **6,389,288 (30)** | **6,389,288 (30)** | 6,389,258 |
| ResNet-152 | 16,232,192 (1,183,990) | **15,048,232 (30)** | **15,048,232 (30)** | 15,048,202 |
| BERT | 28,151,040 (780,480) | **27,569,664 (199,104)** | **27,569,664 (199,104)** | 27,370,560 |

**New Timing Results:**
These are approximate timing results based on my sandbox code. I anticipate the true timing should resemble these numbers. All values are in seconds.

Two processes:
| Model | Run Time - Greedy | Run Time - Sorted-Greedy | Run Time - Multifit |
| --- | --- | --- | --- |
| ResNet-50 | 0.00014 | 0.00014 | 0.00069 |
| ResNet-152 | 0.00036 | 0.00041 | 0.00189 |
| BERT | 0.00015 | 0.00016 | 0.00122 |

Four processes:
| Model | Run Time - Greedy | Run Time - Sorted-Greedy | Run Time - Multifit |
| --- | --- | --- | --- |
| ResNet-50 | 0.00015 | 0.00015 | 0.00095 |
| ResNet-152 | 0.00041 | 0.00042 | 0.00273 |
| BERT | 0.00020 | 0.00019 | 0.00228 |

**Upshot:**
I want to credit the insight from @pritamdamania that `n` in our case should not be too prohibitive in our algorithm design. From the run times, it is certainly feasible to run both greedy and sorted-greedy and take the better partitioning or to run the new algorithm even if may not be able to find a partitioning better than the sorted-greedy algorithm. Either way, the sorted-greedy algorithm appears to be a strong single candidate.
